### ISSUE TITLE
Repeating workflows for different configurations

### ISSUE THREAD

**Issue by DamienIrving** [swcarpentry/make-novice#46]
Once you've got to the point where you are including a configuration file in your makefiles, what is accepted best practice for looping over many different configurations? The only solution I'm aware of is to write a shell script that uses `sed` (inside of a for loop) to alter the relevant text in the configuration file.

This might be something to touch on during the make-novice lessons (even if just in a quick callout box)?

**Comment by mikej888** [comment]
Please could you provide a concrete example or a link, thanks.
I think if I had to use sed to automatically edit a config file, I'd instead be looking at how to change the nature of my configuration to avoid that!

**Comment by DamienIrving** [comment]
Hi @mikej888. The workflow that motivated my question calculates and plots the ocean heat content (OHC) for a given climate model and experiment:
[ohc_base.mk](https://github.com/DamienIrving/climate-analysis/blob/master/workflows/ohc_base.mk)
[ohc_config.mk](https://github.com/DamienIrving/climate-analysis/blob/master/workflows/ohc_config.mk)

I want to run that workflow for a number of different climate models and experiments, so I created a shell script that makes use of `sed` to edit the config file each time around the loop:
[ohc_base.sh](https://github.com/DamienIrving/climate-analysis/blob/master/workflows/ohc_base.sh)

**Comment by mikej888** [comment]
Here is an alternative, which exploits the ability to pass in variable overrides to Make via the command-line. I used a chopped-down version of your ohc_base.mk to test it:

ohc_base.mk:

```
include ohc_config.mk
all : 
	echo ${TARGET}
```

Running this uses the default configuration (with `-s` to specify silent mode):

```
$ make -s -f ohc_base.mk all
/g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r3i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r3i1p3_1956-01-01_2005-12-31.png
```

These can be overridden via values provided at the command line:

```
$ make -s -f ohc_base.mk all ORGANISATION="Edinburgh" MODEL="EPCC" EXPERIMENT=123 RUN=456 VOLUME_RUN=789
/g/data/r87/dbi599/drstree/CMIP5/GCM/Edinburgh/EPCC/123/mon/ocean/ohc-maps/456/ohc-maps_Omon_EPCC_123_456_1956-01-01_2005-12-31.png
```

ohc_base.sh can be changed so that instead of using sed:

```
sed -i "s/^\(ORGANISATION\s*=\s*\).*$/\ORGANISATION=${organisation}/" ohc_config.mk
sed -i "s/^\(MODEL\s*=\s*\).*$/\MODEL=${model}/" ohc_config.mk
sed -i "s/^\(EXPERIMENT\s*=\s*\).*$/EXPERIMENT=${experiment}/" ohc_config.mk
sed -i "s/^\(RUN\s*=\s*\).*$/\RUN=${run}/" ohc_config.mk
sed -i "s/^\(VOLUME_RUN\s*=\s*\).*$/\VOLUME_RUN=${volrun}/" ohc_config.mk
make ${options} -f ohc_base.mk
```

It uses these overrides:

```
make ${options} -f ohc_base.mk ORGANISATION="${organisation}" MODEL="${model}" EXPERIMENT="${experiment}" RUN="${run}" VOLUME_RUN="${volrun}"
```

To check these yield the same results, I captured the output:

```
$ ./ohc_base.sh CSIRO-Mk3-6-0 noAA > original.txt
$ cat original.txt 
echo /g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r1i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r1i1p3_1956-01-01_2005-12-31.png
/g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r1i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r1i1p3_1956-01-01_2005-12-31.png
DONE: CSIRO-Mk3-6-0 historicalMisc r1i1p3: make   -f ohc_base.mk
echo /g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r2i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r2i1p3_1956-01-01_2005-12-31.png
/g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r2i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r2i1p3_1956-01-01_2005-12-31.png
DONE: CSIRO-Mk3-6-0 historicalMisc r2i1p3: make   -f ohc_base.mk
echo /g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r3i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r3i1p3_1956-01-01_2005-12-31.png
/g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historicalMisc/mon/ocean/ohc-maps/r3i1p3/ohc-maps_Omon_CSIRO-Mk3-6-0_historicalMisc_r3i1p3_1956-01-01_2005-12-31.png
DONE: CSIRO-Mk3-6-0 historicalMisc r3i1p3: make   -f ohc_base.mk

$ ./ohc_base_non_sed.sh CSIRO-Mk3-6-0 noAA > non_sed.txt
$ cmp original.txt non_sed.txt 
```

I prefer this as it:

* Avoids sed.
* More importantly, leaves the configuration file intact.

Another option would be to split apart the configuration file into generic configuration and configuration that can be overridden. For example:

ohc_config.1.mk:

```
# zw_climatology_config.mk

# System configuration

PYTHON=/g/data/r87/dbi599/miniconda2/envs/default/bin/python

MY_DATA_DIR=/g/data/r87/dbi599
UA6_DATA_DIR=/g/data/ua6
CMIP5_DIR_START=/drstree/CMIP5/GCM
MY_CMIP5_DIR=${MY_DATA_DIR}${CMIP5_DIR_START}
UA6_CMIP5_DIR=${UA6_DATA_DIR}${CMIP5_DIR_START}

DATA_SCRIPT_DIR=~/climate-analysis/data_processing
VIS_SCRIPT_DIR=~/climate-analysis/visualisation

FIG_TYPE=png
```

ohc_config.2.mk:

```
# Analysis details
ORGANISATION=CSIRO-QCCCE
MODEL=CSIRO-Mk3-6-0
EXPERIMENT=historical
RUN=r1i1p1
VOLUME_RUN=r0i0p0
```

ohc_config.3.mk:

```
CONTROL_RUN=r1i1p1

MAX_DEPTH=2000
START_DATE=1956-01-01
END_DATE=2005-12-31

TARGET=${MY_CMIP5_DIR}/${ORGANISATION}/${MODEL}/${EXPERIMENT}/mon/ocean/ohc-maps/${RUN}/ohc-maps_Omon_${MODEL}_${EXPERIMENT}_${RUN}_${START_DATE}_${END_DATE}.${FIG_TYPE}
```

ohc_base.mk can be defined as:

```
include ohc_config.1.mk
CONFIG=ohc_config.2.mk
include $(CONFIG)
include ohc_config.3.mk

all : 
	echo ${TARGET}
```

The configuration file can then be specified via the command-line e.g.:

ohc_config.custom.mk:

```
# Analysis details
ORGANISATION=Edinburgh
MODEL=EPC
EXPERIMENT=123
RUN=456
VOLUME_RUN=789
```

```
$ make -s -f ohc_base.mk all
/g/data/r87/dbi599/drstree/CMIP5/GCM/CSIRO-QCCCE/CSIRO-Mk3-6-0/historical/mon/ocean/ohc-maps/r1i1p1/ohc-maps_Omon_CSIRO-Mk3-6-0_historical_r1i1p1_1956-01-01_2005-12-31.png

$ make -s -f ohc_base.mk all CONFIG=ohc_config.custom.mk 
/g/data/r87/dbi599/drstree/CMIP5/GCM/Edinburgh/EPC/123/mon/ocean/ohc-maps/456/ohc-maps_Omon_EPC_123_456_1956-01-01_2005-12-31.png
```

ohc_base.sh could be updated so each iteration of the loop programatically build a configuration file which can then be used in place of ohc_config.2.mk. While this results in one configuration file per iteration, the configuration files for each iteration are preserved which may be useful if one iteration fails

**Comment by mikej888** [comment]
I don't think this should be in the lesson proper but part of the Discussion (discussion.md).

**Comment by mikej888** [comment]
Would you like to submit a pull request with changes to discussion.md? (No probs if you don't, I can do it on Monday as part of the bug BBQ)

**Comment by DamienIrving** [comment]
@mikej888 Thanks so much for taking the time to provide an alternative to my workflow. I'm a little tied up with other things for the next few days, but sometime next week I'm happy to submit a pull request to discussion.md.

**Comment by mikej888** [comment]
@DamienIrving, thanks :-)

**Comment by gvwilson** [comment]
any takers?

**Comment by gcapes** [comment]
It looks like this has this been sufficiently addressed in #59? Please let me know if I can close this issue.

**Comment by gcapes** [comment]
In light of the silence, I think #59 fixes this adequately, so I'm closing this issue. Please let me know if you feel differently.
